Dev StreamAnalytics 2017-04-01-preview version for streaming job and 2020-03-01-preview version for dedicated cluster

sunsw1994 opened the pull request:
- Add the 2017-04-01-preview folder and copy the 2016-03-01 files for streaming job related swaggers.
- Add missing features for  2017-04-01-preview 
- Add 2020-03-01-preview version for ASA cluster and private endpoint

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

sunsw1994 committed:
Adds base for updating Microsoft.StreamAnalytics from version stable/2016-03-01 to version 2017-04-01-preview

sunsw1994 committed:
Updates readme

sunsw1994 committed:
Updates API version in new specs and examples

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
1.add missing features for 2017-04-01-preview based on 2016-03-01 2.fix semantic errors

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

sunsw1994 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/common/v1/definitions.json:
```diff
@@ -0,0 +1,198 @@
+        "modelAsString": true
+      }
+    },
+    "Resource": {
```
We need 'Location' in TrackedResource is not required. 
But Location in `commonType#TrackedResource` is required.

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json#L70

weidongxu-microsoft replied:
Interesting. I think ARM requires the location on TrackedResource https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#put-resource

Well, I will leave it to ARM reviewer.

sunsw1994 replied:
The main reason we want to set it no required is when we try to use TrackedResource in the body of Patch Operation, validation will throw error like `property in Patch must not be set required`

weidongxu-microsoft replied:
I see.

Just as reference. Some service would prepare a different Parameter (with only modifiable properties) for PATCH op, e.g. storage service https://github.com/Azure/azure-rest-api-specs/blob/master/specification/storage/resource-manager/Microsoft.Storage/stable/2019-06-01/storage.json#L2794-L2833

sunsw1994 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/operations.json:
```diff
@@ -0,0 +1,124 @@
+    }
+  },
+  "paths": {
+    "/providers/Microsoft.StreamAnalytics/operations": {
```
No, it is just moved from streamingjob.json 

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/stable/2016-03-01/streamingjobs.json#L36

idear1203 replied:
Selected the check box. Thanks for the reminder.

weidongxu-microsoft commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/inputs.json:
```diff
@@ -895,46 +950,102 @@
+        }
+      }
     },
+    "azureSqlReferenceInputDataSource": {
```
These types appears lack description in properties.

sunsw1994 replied:
Thanks for catching it, I will add descriptions for that.

sunsw1994 committed:
add 2020-03-01-preview version for asa cluster and private endpoint

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
address comments 1.resourceGroupName use one unique value in example 2.suppress R3023 validation error 3.add some descriptions

AutorestCI commented:
/azp run automation - sdk

sunsw1994 commented:
For reducing ARM Review Time，this PR includes two version update in two main commit:

- [add missing features for 2017-04-01-preview based on 2016-03-01 ](https://github.com/Azure/azure-rest-api-specs/pull/10673/commits/4ec448438f738f1fa725ee88defdebed71cb5d5e)
   - this part has approved by @weidongxu-microsoft 
   - waiting ARM Review


- [add 2020-03-01-preview version for asa cluster and private endpoint](https://github.com/Azure/azure-rest-api-specs/pull/10673/commits/2fb7c1dd6d0486bc7b8a718d56b8e66ae16ed65c)

weidongxu-microsoft commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2020-03-01-preview/privateEndpoints.json:
```diff
@@ -0,0 +1,358 @@
+      }
+    }
+  },
+  "paths": {
```
Just a reference, some common def https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/privatelinks.json

sunsw1994 replied:
@weidongxu-microsoft  for the reason, share response from @atpham256 at private PR:

> Our PrivateEndpoint definition is different from the common one. We are trying to keep it as close as possible to Azure Network's definition but it is not completely the same.
> 
> Also, for the common definition, your private endpoint definition only contains and "id" property. We need a lot more than that. Customer needs to be able to send us the privateLinkServiceId and groupId, for example.
> 
> Another thing to note here is that we are not a resource like SQL or Storage where a user is connecting to our resource so our integration with private endpoints is not the same. I think Azure Network is using the term "transitive private endpoint access" for our scenario.

https://github.com/Azure/azure-rest-api-specs-pr/pull/1262#discussion_r432212512

majastrz commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/functions.json:
```diff
@@ -0,0 +1,964 @@
+            "$ref": "#/parameters/FunctionNameParameter"
+          }
+        ],
+        "responses": {
```
>responses [](start = 9, length = 9)

You need to declare a default error response in all the operations in all the swaggers in this PR.

sunsw1994 replied:
Do you mean that we need to add a default error response like:
```
        "responses": {
          "200": {

          },
          "202": {

          },
          "default": {
            "description": "Error.",
            "schema": {
              "$ref": "../../common/v1/definitions.json#/definitions/Error"
            }
          }
        },
```


https://github.com/Azure/azure-rest-api-specs/blob/aa91867b7b83be0534ce3feb2583c12e0cdd2bff/specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2020-03-01-preview/clusters.json#L90

majastrz reviewed (approved):
Signed off from ARM side.

weidongxu-microsoft commented:
@sunsw1994
All approved.
Let me know when you need it merged.

sunsw1994 commented:
> @sunsw1994
> All approved.
> Let me know when you need it merged.

@weidongxu-microsoft thank you very much, I will update code based on ARM team's and your comments. And will let you need when it is ready to be merged. 

Another question, how do we update the documentation of REST API published at https://docs.microsoft.com/en-us/rest/api/azure/ ? Could you share some guides for it? Thanks.

weidongxu-microsoft commented:
@sunsw1994

For publishing doc, please follow https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/79/Generation-of-docs-on-docs.microsoft.com

changlong-liu commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/examples/Function_Delete.json:
```diff
@@ -0,0 +1,13 @@
+  "parameters": {
+    "api-version": "2017-04-01-preview",
+    "subscriptionId": "56b5e0a9-b645-407d-99b0-c64f86013e3d",
+    "resourceGroupName": "sjrg1637",
```
Is there special reason on using diferent resourceGroupName on different examples?
To make the examples available for generating test, they should use one unique resource group

sunsw1994 replied:
Hi @changlong-liu  thank you for bring it up, I think there are no special reason on using different resourceGroupName. That may just be copied from real request/response. 

I get context from email thread that you are working on helping ASA to auto-generate CLI based on swagger files, right? 

Do we just need to update  resourceGroupName to using one unique value? Are there any other issues may block generating CLI?

sunsw1994 replied:
Updated

changlong-liu commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2020-03-01-preview/clusters.json:
```diff
@@ -0,0 +1,612 @@
+      ],
+      "properties": {
+        "sku": {
+          "$ref": "#/definitions/ClusterSku"
```
in the live validation, I found sku is a required parameter in service side, Please consider whether need to add "required" in swagger

sunsw1994 replied:
@atpham256 do we need to set `sku` as a required parameter? What happens if customer try to send a create/update request without `sku` property. @changlong-liu is helping to do live validation for CLI

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
Merge branch 'master' into dev-streamanalytics-Microsoft.StreamAnalytics-2017-04-01-preview

AutorestCI commented:
/azp run automation - sdk

atpham256 commented on specification/streamanalytics/resource-manager/readme.md:
```diff
@@ -4,33 +4,80 @@
+
+These settings apply only when `--tag=package-2020-03-preview` is specified on the command line.
+
+This section contains the composite set of APIs. Deticate cluster and private endpoint related APIs are under 2020-03-01-preview version. Streamingjobs related APIs are under 2017-04-01-preview version.
```
Typo.

However, we should refer to it as Stream Analytics Cluster and Private Endpoint as that is what we are advertising. We aren't using the term "dedicated".

atpham256 replied:
This comment applies to multiple places

sunsw1994 replied:
Thanks for catching that. Fixed.

atpham256 commented on specification/streamanalytics/resource-manager/readme.md:
```diff
@@ -4,33 +4,80 @@
+
+These settings apply only when `--tag=package-pure-2020-03-preview` is specified on the command line.
+
+This section only contains input swagger files for version 2020-03-01-preview, incluing deticate cluster and private endpoint related APIs.  
```
including

I see this typo in multiple places as well

sunsw1994 replied:
Fixed

atpham256 commented on specification/streamanalytics/resource-manager/readme.md:
````diff
@@ -4,33 +4,80 @@
+  - Microsoft.StreamAnalytics/preview/2020-03-01-preview/privateEndpoints.json
+```
+
+### Tag: package-2017-04-preview
````
shouldn't this be pure?

sunsw1994 replied:
updated.

atpham256 commented on specification/streamanalytics/resource-manager/readme.md:
````diff
@@ -4,33 +4,80 @@
+  - Microsoft.StreamAnalytics/preview/2017-04-01-preview/operations.json
+```
 
 ### Tag: package-2016-03
````
should this be pure?

sunsw1994 replied:
updated.

atpham256 commented on specification/streamanalytics/resource-manager/readme.go.md:
````diff
@@ -24,3 +26,20 @@ Please also specify `--go-sdk-folder=<path to the root directory of your azure-s
 output-folder: $(go-sdk-folder)/services/$(namespace)/mgmt/2016-03-01/$(namespace)
 ```
+
+### Tag: package-2017-04-preview and go
````
Is there no pure vs composite distinction in Go?

atpham256 commented on specification/streamanalytics/resource-manager/readme.md:
````diff
@@ -123,6 +168,31 @@ regenerate-manager: true
+generate-interface: true
+```
+
+### Tag: package-2020-03-preview and java
````
Is there no distinction between pure vs composite in Java?

atpham256 replied:
Similar question for the other languages

sunsw1994 replied:
> Is there no distinction between pure vs composite in Java?

Thank you for catching it.
There may be only one Tag to represent 2020_03_01_preview in Java or Go, `pure` or `composite`. Because they share the same namespace and path name. We think using 'package-pure-2020-03-preview'  is better here, I will update that.

For reference:

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/readme.md#tag-package-compute-only-2017-12-and-java

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/readme.go.md#tag-package-compute-2017-12-and-go

atpham256 replied:
What about the other languages?

How will SDK generation work for languages like Java or Go then if we are only doing the pure version of the API? Are you saying users using Java SDK will only be able to call cluster APIs but not streaming job APIs?

sunsw1994 replied:
For my understand,  Java will generate SDK based on :
```
### Java multi-api

``` yaml $(java) && $(multiapi)
batch:
  - tag: package-pure-2017-04-preview
  - tag: package-pure-2020-03-preview
```
that will include pure-2017-04-preview and pure-2020-03-preview, released package name may be like `azure-mgmt-streamanalytics`. Users using Java SDK will be able to call both cluster APIs and streaming job APIs?

***

For Go, SDK will be published in Github, and sdk code path will include api version,like:
```
azure-sdk-for-go/services/streamanalytics/mgmt/2016-03-01/streamanalytics/
azure-sdk-for-go/services/preview/streamanalytics/mgmt/2017-04-01-preview/streamanalytics/
azure-sdk-for-go/services/preview/streamanalytics/mgmt/2020-03-01-preview/streamanalytics/
```
Users using go SDK will be able to call streaming job api by importing .../mgmt/2017-04-01-preview/streamanalytics, call cluster api by importing ...mgmt/2020-03-01-preview/streamanalytics/

reference：
https://github.com/Azure/azure-sdk-for-go/tree/master/services/streamanalytics/mgmt/2016-03-01/streamanalytics

***

For ruby, I don't know about that, looks closer to java

***

For C# and ts, using tag=package-2020-03-preview, published SDK will include composite api version.

***
We can confirm that with @weidongxu-microsoft  He may give us the best practices

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/inputs.json:
```diff
@@ -883,46 +950,102 @@
+        }
+      }
     },
+    "azureSqlReferenceInputDataSource": {
```
I think generally definitions should start with a capital letter. AzureSqlReferenceInputDataSource

The properties that will be in the JSON are generally camelCase

sunsw1994 replied:
Fixed.

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/operations.json:
```diff
@@ -0,0 +1,124 @@
+    }
+  },
+  "paths": {
+    "/providers/Microsoft.StreamAnalytics/operations": {
```
I thought you moved this to a separate file?

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/streamingjobs.json:
```diff
@@ -563,6 +549,10 @@
           "type": "string",
           "description": "Describes the state of the streaming job."
         },
+        "jobType": {
```
Why is this not an enum string?

sunsw1994 replied:
Updated

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/streamingjobs.json:
```diff
@@ -639,10 +629,38 @@
+          "$ref": "#/definitions/External",
+          "description": "External Property."
+        },
+        "cluster": {
```
this is an object

atpham256 replied:
Example of how it will look in the json:

"cluster": { 
   "id": "/subscriptions/409468b9-1047-4f2c-b98b-9245c3d1448d/resourceGroups/IoTResources/providers/Microsoft.StreamAnalytics/clusters/MyCluster" 
},

sunsw1994 replied:
Fixed

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/inputs.json:
```diff
@@ -461,7 +471,10 @@
         }
-      }
+      },
+      "required": [
```
Why do you need this "required" annotation?

atpham256 replied:
I think I also see this for other "type" like properties across the files

sunsw1994 replied:
For fix SEMANTIC_VALIDATION_ERROR, it may be a new CI rule, so before code don't throw such error when checked in.
```
{
  code: 'SEMANTIC_VALIDATION_ERROR',
  id: 'OAV119',
  message: 'The spec .\\specification\\streamanalytics\\resource-manager\\Microsoft.StreamAnalytics\\preview\\2017-04-01-preview\\inputs.json has semantic validation errors.',
  innerErrors: [
    {
      code: 'OAV131',
      error: 'DISCRIMINATOR_NOT_REQUIRED',
      name: 'discriminator',
      params: [ 'type' ],
      message: 'discriminator must be a required property.',
      path: [ 'definitions,Serialization' ]
    }
  ],
  level: '\u001b[31merror\u001b[39m'
}
```

atpham256 commented:
I don't see examples for the new properties/feature you added. You should add them not only for documentation but to also validate the swagger based on example JSONs.

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2020-03-01-preview/clusters.json:
```diff
@@ -0,0 +1,612 @@
+      ],
+      "x-ms-enum": {
+        "name": "JobState",
+        "modelAsString": false,
```
I think this should be modelAsString = true to be flexible/backwards compatible unless there is a good reason to be false here

sunsw1994 replied:
fixed

sunsw1994 committed:
Merge branch 'master' into dev-streamanalytics-Microsoft.StreamAnalytics-2017-04-01-preview

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
address comments:
1. update readme
2. modify cluster type:string to object
3. update difinition of AzureSqlReferenceInputDataSource

AutorestCI commented:
/azp run automation - sdk

sunsw1994 commented:
@weidongxu-microsoft  could you help us to re-review the latest readme files for each language? We have tow new api versions, `but these contains different features and no intersection`.

Which is the better or standard way for Java/Go/Ruby:

1. Put 2020-03-01-preview and 2017-04-01-preview into a composite package. Using 2020-03-01-preview as package suffix.
2. Separate packages for 2020-03-01-preview and 2017-04-01-preview. 

Thank you very much.

atpham256 commented:
> I don't see examples for the new properties/feature you added. You should add them not only for documentation but to also validate the swagger based on example JSONs.

Ping on this comment. It is important to have examples not only for documentation but also to validate the swagger. This will make it more likely that the SDK will work as expected.

sunsw1994 committed:
update example json files

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
add readme for python

AutorestCI commented:
/azp run automation - sdk

sunsw1994 committed:
update readme for go

AutorestCI commented:
/azp run automation - sdk

sunsw1994 commented:
> > I don't see examples for the new properties/feature you added. You should add them not only for documentation but to also validate the swagger based on example JSONs.
> 
> Ping on this comment. It is important to have examples not only for documentation but also to validate the swagger. This will make it more likely that the SDK will work as expected.

Have updated part of example json files at [commit](https://github.com/Azure/azure-rest-api-specs/pull/10673/commits/58c3e81aa7e1b347e218457d07d6ec567a5b9f8f)

atpham256 commented on specification/streamanalytics/resource-manager/Microsoft.StreamAnalytics/preview/2017-04-01-preview/examples/Output_Create_PowerBI.json:
```diff
@@ -0,0 +1,76 @@
+              "tokenUserDisplayName": "Bob Smith",
+              "groupId": "ac40305e-3e8d-43ac-8161-c33799f43e95",
+              "groupName": "MyPowerBIGroup",
+              "authenticationMode": "ConnectionString"
```
You should make sure examples have reasonable values. ConnectionString doesn't make sense for a PowerBI output
